MudSwitch: Add Size Parameter

SinisterMaya opened the pull request:
<!--
MAKE SURE TO READ THE CONTRIBUTING GUIDE BEFORE CREATING A PR
https://github.com/MudBlazor/MudBlazor/blob/dev/CONTRIBUTING.md

Testing sections can be removed for documentation changes
-->

<!-- Provide a general summary of your changes in the Title above -->
<!-- Keep the title short and descriptive, as it will be used as a commit message -->


## Description
<!-- Describe your changes in detail any why -->
<!-- Note any issues that are resolved by this PR -->
<!-- e.g. resolves #1337 or fixes #9310 -->

Adds parameter Size to the MudSwitch component so it can be smaller or larger. 

- Changed the SCSS to include classes for small, medium and large switch
- Modified the Razor page so it uses CSSBuilder for css classes instead of being hardcoded
- Made the CSSBuilder add size related Classes by using the Size Parameter
- Changed the Razor ThumbIcon Style to use a combination of padding and height/width inheritance so the size of the icon is always 4px less than the thumb itself (Original was 20px for thumb and 16px for Icon). That way, original appearance remains.
- Parameter has default value set to medium so original apprearance remains.



## How Has This Been Tested?
<!-- All PR's should implement unit tests if possible -->
<!-- Please describe how you tested your changes. -->
<!-- Have you created new tests or updated existing ones? -->
<!-- e.g. unit | visually | none -->

All three sizes have been added in the unit test viewer

![image](https://user-images.githubusercontent.com/29356866/200126781-0aefe400-eb5f-4c4e-bf43-742293240d47.png)
![image](https://user-images.githubusercontent.com/29356866/200127019-78cb5294-51fd-4a1b-a27c-096c319a5436.png)

## Types of changes
<!-- What types of changes does your code introduce? Put an `x` in all the boxes that apply: -->
- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)

<!-- If you made any visual changes, provide screenshots of before/after, it its moving parts, please provide high quality gif, wemb or mp4 -->

## Checklist:
<!-- Go over all the following points, and put an `x` in all the boxes that apply. -->
<!-- If you're unsure about any of these, don't hesitate to ask. We're here to help! -->
- [x] The PR is submitted to the correct branch (`dev`).
- [x] My code follows the code style of this project.
- [x] I've added relevant tests.

rogue6800 commented:
This looks great! I would be using the small option all the time in dense forms!

TDroogers commented:
looks great, but shouldn't the label also be adjusted?

SinisterMaya commented:
About the label size, I thought about that, but figured that since label are sized using a MudText Component, I wasn't sure of what is the "standard" way of sizing in that case. Do we make it Typo.h6 for large and typo.body2 for small ? has this been used in a similar cased before ?

In the case of MudButton, they override, font-size to 0.8125rem for small and 0.9375rem for large while medium uses typography-body2 which is at 0.875rem. should we also make it ± 0.0625 rem ?

MS-Hayden commented:
How about supplying the Typo as a parameter?

snakex64 commented:
> How about supplying the Typo as a parameter?

For more advanced users it might be useful, but I think it would easier to just change the size of the font internally based on the size of the switch. Worst case a flag to disable the scaling of the text could be added, even though it seems logical to scale the text too.

SinisterMaya commented:
EDIT:

Added Control for label/text size

![image](https://user-images.githubusercontent.com/29356866/200682549-f20d554f-500e-4325-82ae-29097698adba.png)
![image](https://user-images.githubusercontent.com/29356866/200682593-369af94f-48b2-40c8-a162-7c94759dcaac.png)

Added Parameter LabelTypo that provides control over the Typo Parameter of the MudText component used for the label.

Added a check if the LabelTypo is explicitly set on the switch. If yes, the label will use this parameter for its typography. Otherwise, the typography will be infered from the Size parameter. 

Original appearance is maintained by using the default Typo.Body1 for the default Size.Medium value

mikes-gh commented:
@SinisterMaya Great work 😄  One last thing could you write test for the new parameters 🙏

mikes-gh commented:
@JonBunator What do you think

JonBunator reviewed (commented):
Thanks for the PR. I like it. But I think we should debate the changeability of the typo parameter. We should stay consistent across the library. If we add it for switches, we should also add it to all other components: tabs, buttons, checkboxes etc. And I am not sure if that makes sense.
So, I think for now the same approach that MudButton uses with its different sizes would be better for now

SinisterMaya commented:
Ok, Should I go ahead and change the LabelTypo property for a private field then ?

mikes-gh commented:
> Ok, Should I go ahead and change the LabelTypo property for a private field then ?

IMO yes `_labelTypo` for now because then we can get this merged and continue the discussion about Typo manual overrides in the context of the whole library.

SinisterMaya commented:
EDIT:

Removed the LabelTypo parameter, changed it to a private field so it's no longer something the user has control over. 

Modified my unit test to validate that Size Parameter acts on the label size even when changed dynamically.

![image](https://user-images.githubusercontent.com/29356866/201547557-16c35c6a-5d43-48ae-bc1c-10b4b8c81a7a.png)
![image](https://user-images.githubusercontent.com/29356866/201547559-ad2f84dc-cde5-4e32-bb65-284731c31eee.png)

mikes-gh commented:
@SinisterMaya not sure I like the Heading Body fonts as labels.  Just doesn't look right IMO, especially since the boldness changes.  I think you also had a feeling like this due to your previous comments.

SinisterMaya commented:
> @SinisterMaya not sure I like the Heading Body fonts as labels. Just doesn't look right IMO, especially since the boldness changes. I think you also had a feeling like this due to your previous comments.

You are Right. To me, me most important is the small size, so we can use switches in datagrids and in other situations where you want a denser component. I personally don't like the button approach. It simply scales its font size. I prefer having direct control onto my fonts instead of it being scaled from another one. Since the rest of the switch size is not based on font size and is hard-coded. Maybe we should just have the font size of its label overridden based on a line height value ? Make it so, we keep the font family from body 1, but make the size fixed for a consistent end result ?

henon commented:
This looks like a good PR which got held up by a indecision on typography. Can we make it consistent with other components that are sizable? And who doesn't like how that looks can override with CSS or set the text to null and put their own MudText beside?

SinisterMaya commented:
Alright, I am gonna do complete this soon and resubmit

Commit:
MudSwitch: Added size parameter to haver a smaller or larger switch

SinisterMaya commented:
EDIT:

In order to reproduce button component behavior, I added 3 css classes for switch label font-size since the css property has to be on the <p> level otherwise the default typography parameter of MudText component prevails. Those classes are applied to MudText component of MudSwitch from a CssBuilder that uses the Size parameter.

Since Button are not scaled the same as regular component (default is not 1rem) I did the following: 

- small: since small button is at 0.8125rem (13/16), I did the same 
- medium: I gave the default size of actual switch which is 1rem (Typo.Body1)
- large: I gave the same offset as for small so 1.1875rem (1 3/16)

The result gives 13px for small, 16px for medium and 19px for large:
![image](https://user-images.githubusercontent.com/29356866/217082573-f09f5c85-f7f0-46e4-85e2-083b364c503b.png)
![image](https://user-images.githubusercontent.com/29356866/217082617-b4caa507-0e3c-4650-9080-0206b4264ff1.png)

henon commented:
Thanks!

Yomodo commented:
Awesome PR! The doc is a bit messy now; whenever I toggle the Success, the other switches also get switched:

![image](https://user-images.githubusercontent.com/10358198/217525410-54ecf22d-0318-4767-bbb5-9982acda93cd.png)

![image](https://user-images.githubusercontent.com/10358198/217525471-3b3c6122-02ea-4221-890e-b04c177f9a26.png)

SinisterMaya commented:
@Yomodo Yeah, I linked them all to the same bool to ease testing... I guess i'll do another PR for the Docs in order to add the Size example and clean that up ?

Yomodo commented:
> I guess i'll do another PR for the Docs in order to add the Size example and clean that up ?

If you please, to avoid any confusion or future doc-bug reports.
